File: Python/Circular Linked List.py
class Node:
    def __init__(self, data):
       self.data, self.next = data, None
 
class CircularlyLinkedList:
    def __init__(self):
        self.head, self.tail, self.nodes = None, None, 0
    
    def insertStart(self, data):
        new = Node(data)
        if self.head is None:
            self.head, self.tail = new, new
        else:
            new.next = self.head
            self.head = new
            self.tail.next = new
        self.nodes += 1
    
    def insertMid(self, pos, data):
        if self.head is None:
            new = Node(data)
            self.head, self.tail = new, new
            self.nodes += 1
        else:
            if pos == 1:
                self.insertStart(data)
            elif pos >= self.nodes:
                self.insertEnd(data)
            else:
                new, temp = Node(data), self.head
                for _ in range(pos-2):
                    temp = temp.next
                new.next = temp.next
                temp.next = new
                self.nodes += 1
    
    def insertEnd(self, data):
        new = Node(data)
        if self.head is None:
            self.head, self.tail = new, new
        else:
            self.tail.next = new
            self.tail = new
            new.next = self.head
        self.nodes += 1
    
    def deleteStart(self):
        if self.head is None:
            return None
        else:
            temp = self.head
            self.head = self.head.next
            self.tail.next = self.head
            temp.next = None
            self.nodes -= 1
            return temp
    
    def deleteMid(self, pos):
        if self.head is None:
            return None
        else:
            if pos == 1:
                self.deleteStart()
            elif pos >= self.nodes:
                self.deleteEnd()
            else:
                curr, temp = self.head, self.head.next
                for i in range(pos-2):
                    curr, temp = curr.next, temp.next
                curr.next = temp.next
                temp.next = None
                self.nodes -= 1
                return temp
    
    def deleteEnd(self):
        if self.head is None:
            return None
        else:
            curr = self.head
            for i in range(self.nodes-2):
                curr = curr.next
            temp = curr.next
            curr.next = self.head
            self.tail = curr
            temp.next = None
            self.nodes -= 1
            return temp
    
    def display(self):
        temp = self.head
        for i in range(self.nodes):
            print(temp.data, end='-->')
            temp = temp.next
        print()

File: JavaScript/Circular Linked List.js
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class CircularLinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.nodes = 0;
    }
    insertStart(data) {
        const newNode = new Node(data);
        if (this.head === null) {
            this.head = newNode;
            this.tail = newNode;
        }
        else {
            newNode.next = this.head;
            this.head = newNode;
            this.tail.next = this.head;
        }
        this.nodes += 1;
    }
    insertMid(pos, data) {
        if (this.head === null) {
            const newNode = new Node(data);
            this.head = newNode;
            this.tail = newNode;
            this.nodes += 1;
        }
        else {
            if (pos == 1)
                this.insertStart(data);
            else if (pos >= this.nodes)
                this.insertEnd(data);
            else {
                const newNode = new Node(data);
                let temp = this.head;
                for (let i=0; i<pos-2; i++)
                    temp = temp.next;
                newNode.next = temp.next;
                temp.next = newNode;
                this.nodes += 1;
            }
        }
    }
    insertEnd(data) {
        const newNode = new Node(data);
        if (this.head === null) {
            this.head = newNode;
            this.tail = newNode;
        }
        else {
            this.tail.next = newNode;
            this.tail = newNode;
            this.tail.next = this.head;
        }
        this.nodes += 1;
    }
    deleteStart() {
        if (this.head === null)
            console.log('Empty Circular Linked List');
        else {
            const temp = this.head;
            this.head = this.head.next;
            this.tail.next = this.head;
            temp.next = null;
            this.nodes -= 1;
        }
    }
    deleteMid(pos) {
        if (this.head === null)
            console.log('Empty Circular Linked List');
        else {
            if (pos == 1)
                this.deleteStart();
            else if (pos >= this.nodes)
                this.deleteEnd();
            else {
                let curr = this.head, temp = this.head.next;
                for (let i=0; i<pos-2; i++) {
                    curr = curr.next;
                    temp = temp.next;
                }
                curr.next = temp.next;
                temp.next = null;
                this.nodes -= 1;
            }
        }
    }
    deleteEnd() {
        if (this.head === null)
            console.log('Empty Circular Linked List');
        else {
            let temp = this.head;
            for (let i=0; i<this.nodes-2; i++) {
                temp = temp.next;
            }
            temp.next = this.head;
            this.tail.next = null;
            this.tail = temp;
            this.nodes -= 1;
        }
    }
    display() {
        if (this.head === null)
            console.log('Empty Circular Linked List');
        else {
            let temp = this.head, data = ``;
            for (let i=0; i<this.nodes; i++) {
                data += `${temp.data}->`
                temp = temp.next;
            }
            console.log(data);
        }
    }
}
